File: src/lib/analysis/realized-pnl.ts
import type { Trade } from "@/types/database";

export type TradeGroupKey = {
  ticker: string | null;
  assetName: string;
  country: string;
  accountId: string;
};

export type MutationType = "insert" | "update" | "delete";

export type Mutation =
  | { type: "insert"; trade: Trade }
  | { type: "update"; trade: Trade; patch: Partial<Trade> }
  | { type: "delete"; trade: Trade };

export type GroupPnLEntry = {
  profit_loss: number;
  avg_buy_price: number;
  matched_qty: number;
  running_qty_after: number;
};

export type ValidateMutationResult =
  | { ok: false; message: string }
  | { ok: true; affectedSellIds: string[]; newPnL: Map<string, number> };

// (종목+국가+계좌) 그룹 키 문자열
export function groupKey(trade: Pick<Trade, "ticker_symbol" | "asset_name" | "country_code" | "account_id">): string {
  return `${trade.ticker_symbol ?? trade.asset_name}:${trade.country_code ?? "KR"}:${trade.account_id}`;
}

// 그룹 매칭 — ticker_symbol(없으면 asset_name) 기준 단일 비교
// migration 006에서 모든 레코드의 ticker_symbol이 보장됨
function isSameGroup(trade: Trade, key: TradeGroupKey): boolean {
  if (trade.account_id !== key.accountId) return false;
  if ((trade.country_code ?? "KR") !== key.country) return false;
  const tradeTicker = trade.ticker_symbol ?? trade.asset_name;
  const targetTicker = key.ticker ?? key.assetName;
  return tradeTicker === targetTicker;
}

// 계산용 정렬: traded_at asc → 같은 날이면 BUY 먼저 → created_at asc
export function sortForCalc(trades: Trade[]): Trade[] {
  return [...trades].sort((a, b) => {
    const tDiff = new Date(a.traded_at).getTime() - new Date(b.traded_at).getTime();
    if (tDiff !== 0) return tDiff;
    if (a.trade_type !== b.trade_type) return a.trade_type === "BUY" ? -1 : 1;
    return new Date(a.created_at).getTime() - new Date(b.created_at).getTime();
  });
}

// WAC 기반 SELL 손익 계산 (항상 계산값 사용 — 수동 입력 불가)
export function sellPnL(trade: Trade, avgCost: number, costQty?: number): number {
  const qty = costQty ?? trade.quantity;
  return trade.price * qty - avgCost * qty - trade.commission - trade.tax;
}

// 특정 그룹의 SELL별 profit_loss 계산 — Map<sellId, GroupPnLEntry>
export function computeGroupPnL(
  trades: Trade[],
  key: TradeGroupKey,
): Map<string, GroupPnLEntry> {
  const result = new Map<string, GroupPnLEntry>();

  const group = sortForCalc(trades.filter((t) => isSameGroup(t, key)));

  let runningQty = 0;
  let runningCost = 0;

  for (const trade of group) {
    if (trade.trade_type === "BUY") {
      runningQty += trade.quantity;
      runningCost += trade.price * trade.quantity;
    } else {
      const avgCost = runningQty > 0 ? runningCost / runningQty : 0;
      const matchedQty = Math.min(trade.quantity, runningQty);
      result.set(trade.id, {
        profit_loss: sellPnL(trade, avgCost, matchedQty),
        avg_buy_price: avgCost,
        matched_qty: matchedQty,
        running_qty_after: Math.max(0, runningQty - trade.quantity),
      });
      runningCost = Math.max(0, runningCost - avgCost * matchedQty);
      runningQty = Math.max(0, runningQty - trade.quantity);
    }
  }

  return result;
}

// 수정/삭제/삽입 가상 적용 후 oversell 여부 검증
export function validateMutation(trades: Trade[], mutation: Mutation): ValidateMutationResult {
  let virtual: Trade[];
  const mutTrade = mutation.type === "update" ? mutation.trade : mutation.trade;
  const key: TradeGroupKey = {
    ticker: mutTrade.ticker_symbol,
    assetName: mutTrade.asset_name,
    country: mutTrade.country_code ?? "KR",
    accountId: mutTrade.account_id,
  };

  if (mutation.type === "insert") {
    virtual = [...trades, mutation.trade];
  } else if (mutation.type === "update") {
    const patched = { ...mutation.trade, ...mutation.patch };
    virtual = trades.map((t) => (t.id === mutation.trade.id ? patched : t));
  } else {
    virtual = trades.filter((t) => t.id !== mutation.trade.id);
  }

  const group = sortForCalc(virtual.filter((t) => isSameGroup(t, key)));

  let runningQty = 0;
  let runningCost = 0;
  const affectedSellIds: string[] = [];
  const newPnL = new Map<string, number>();

  for (const trade of group) {
    if (trade.trade_type === "BUY") {
      runningQty += trade.quantity;
      runningCost += trade.price * trade.quantity;
    } else {
      if (runningQty <= 0) {
        return { ok: false, message: "보유 수량이 없어 매도할 수 없습니다." };
      }
      if (trade.quantity > runningQty) {
        return { ok: false, message: "보유 수량이 부족한 매도 거래가 생깁니다." };
      }
      const avgCost = runningCost / runningQty;
      const matchedQty = Math.min(trade.quantity, runningQty);
      const pnl = sellPnL(trade, avgCost, matchedQty);
      affectedSellIds.push(trade.id);
      newPnL.set(trade.id, pnl);
      runningCost = Math.max(0, runningCost - avgCost * matchedQty);
      runningQty = Math.max(0, runningQty - trade.quantity);
    }
  }

  return { ok: true, affectedSellIds, newPnL };
}

// SELL 거래의 P&L 반환 — 저장값 우선, null이면 fallbackMap 조회 (백필 전 호환)
export function getPnL(trade: Trade, fallbackMap?: Map<string, number>): number {
  if (trade.profit_loss != null) return Number(trade.profit_loss);
  return fallbackMap?.get(trade.id) ?? 0;
}

// 저장된 profit_loss 우선 pnlMap 생성 (분석/집계 전달용)
// null인 거래는 WAC fallback으로 채움 (백필 완료 후에는 항상 저장값 사용)
export function buildPnlMap(trades: Trade[]): Map<string, number> {
  const fallback = computeRealizedPnL(trades);
  const result = new Map<string, number>();
  for (const t of trades) {
    if (t.trade_type === "SELL") {
      result.set(t.id, t.profit_loss != null ? Number(t.profit_loss) : (fallback.get(t.id) ?? 0));
    }
  }
  return result;
}

// 전체 거래에서 각 SELL trade.id → 실현손익 (fallback 경로용)
export function computeRealizedPnL(trades: Trade[]): Map<string, number> {
  const result = new Map<string, number>();
  const sorted = sortForCalc(trades);
  const posMap = new Map<string, { runningQty: number; runningCost: number }>();

  for (const trade of sorted) {
    const key = groupKey(trade);
    if (!posMap.has(key)) posMap.set(key, { runningQty: 0, runningCost: 0 });
    const pos = posMap.get(key)!;

    if (trade.trade_type === "BUY") {
      pos.runningQty += trade.quantity;
      pos.runningCost += trade.price * trade.quantity;
    } else {
      const avgCost = pos.runningQty > 0 ? pos.runningCost / pos.runningQty : 0;
      const matchedQty = Math.min(trade.quantity, pos.runningQty);
      result.set(trade.id, sellPnL(trade, avgCost, matchedQty));
      pos.runningCost = Math.max(0, pos.runningCost - avgCost * matchedQty);
      pos.runningQty = Math.max(0, pos.runningQty - trade.quantity);
    }
  }

  return result;
}

File: api/src/invest_note_api/domain/realized_pnl.py
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
import math
from typing import Literal

from invest_note_api.domain.trade_types import (
    RESULT_BREAKEVEN,
    RESULT_FAIL,
    RESULT_SUCCESS,
    STRATEGY_UNKNOWN,
    TRADE_TYPE_BUY,
    TRADE_TYPE_SELL,
    EmotionType,
    ReasoningTag,
    StrategyType,
    Trade,
    TradeResult,
    trade_country,
    trade_identifier,
)
from invest_note_api.domain.trade_utils import MS_PER_DAY, to_kst_ms
from invest_note_api.domain.trade_walker import (
    ConsumedLot,
    walk_trades,
)


@dataclass(frozen=True)
class TradeGroupKey:
    ticker: str | None
    asset_name: str
    country: str
    account_id: str


MutationType = Literal["insert", "update", "delete"]


def trade_to_group_key(trade: Trade) -> TradeGroupKey:
    return TradeGroupKey(
        ticker=trade.ticker_symbol,
        asset_name=trade.asset_name,
        country=trade_country(trade),
        account_id=trade.account_id,
    )


def is_same_group(trade: Trade, key: TradeGroupKey) -> bool:
    if trade.account_id != key.account_id:
        return False
    if trade_country(trade) != key.country:
        return False
    trade_ticker = trade_identifier(trade)
    target_ticker = key.ticker or key.asset_name
    return trade_ticker == target_ticker


def sort_for_calc(trades: list[Trade]) -> list[Trade]:
    """traded_at 오름차순, 동시각은 BUY 먼저, 그 다음 created_at."""
    return sorted(
        trades,
        key=lambda t: (
            t.traded_at,
            0 if t.trade_type == TRADE_TYPE_BUY else 1,
            t.created_at,
        ),
    )


def _sell_pnl(trade: Trade, avg_cost: float, cost_qty: float | None = None) -> float:
    qty = cost_qty if cost_qty is not None else trade.quantity
    return trade.price * qty - avg_cost * qty - trade.commission - trade.tax


@dataclass
class GroupPnLEntry:
    profit_loss: float
    avg_buy_price: float
    holding_days: int | None
    strategy_type: StrategyType | None
    reasoning_tags: list[ReasoningTag]
    emotion: EmotionType | None
    result: TradeResult
    matched_qty: float
    running_qty_after: float


def derive_result_from_pnl(pnl: float) -> TradeResult:
    if pnl > 0:
        return RESULT_SUCCESS
    if pnl < 0:
        return RESULT_FAIL
    return RESULT_BREAKEVEN


def _strategy_from_consumed(consumed: Sequence[ConsumedLot]) -> StrategyType | None:
    if not consumed:
        return None

    by_strategy: dict[str, dict] = {}
    for c in consumed:
        key = c.lot.strategy or STRATEGY_UNKNOWN
        if key not in by_strategy:
            by_strategy[key] = {"qty": 0.0, "order": c.lot.order}
        by_strategy[key]["qty"] += c.qty
        by_strategy[key]["order"] = min(by_strategy[key]["order"], c.lot.order)

    selected = sorted(by_strategy.items(), key=lambda item: (-item[1]["qty"], item[1]["order"]))[0][0]
    return selected  # type: ignore[return-value]


def _meta_from_consumed_latest(
    consumed: Sequence[ConsumedLot],
) -> tuple[list[ReasoningTag], EmotionType | None]:
    """소비된 BUY lot 중 가장 최근(time_ms 최대, 동률 시 order 최대)의 tags/emotion."""
    if not consumed:
        return [], None
    latest = max(consumed, key=lambda c: (c.lot.time_ms, c.lot.order))
    return list(latest.lot.reasoning_tags), latest.lot.emotion


def _holding_days_from_consumed(
    consumed: Sequence[ConsumedLot], sell_time_ms: int
) -> int | None:
    total = sum(c.qty for c in consumed)
    if total <= 0:
        return None
    weighted_ms = sum((sell_time_ms - c.lot.time_ms) * c.qty for c in consumed)
    return math.floor(weighted_ms / total / MS_PER_DAY + 0.5)


def compute_group_pnl(trades: list[Trade], key: TradeGroupKey) -> dict[str, GroupPnLEntry]:
    """그룹 내 SELL 거래별 WAC PnL 계산."""
    result: dict[str, GroupPnLEntry] = {}

    for ev in walk_trades(
        trades,
        group_filter=lambda t: is_same_group(t, key),
        sort_fn=sort_for_calc,
    ):
        if ev.kind != "SELL":
            continue

        sell_time_ms = to_kst_ms(ev.trade.traded_at)
        avg_cost = ev.state_before.avg_cost
        pnl = _sell_pnl(ev.trade, avg_cost, ev.matched_qty)
        tags, emotion = _meta_from_consumed_latest(ev.consumed)

        result[ev.trade.id] = GroupPnLEntry(
            profit_loss=pnl,
            avg_buy_price=avg_cost,
            holding_days=_holding_days_from_consumed(ev.consumed, sell_time_ms),
            strategy_type=_strategy_from_consumed(ev.consumed),
            reasoning_tags=tags,
            emotion=emotion,
            result=derive_result_from_pnl(pnl),
            matched_qty=ev.matched_qty,
            running_qty_after=ev.state_after.running_qty,
        )

    return result


def _apply_virtual(
    trades: list[Trade],
    mutation_type: MutationType,
    trade: Trade,
    patch: dict | None,
) -> list[Trade]:
    if mutation_type == "insert":
        return [*trades, trade]
    if mutation_type == "update":
        patched_data = {**trade.model_dump(), **(patch or {})}
        patched = Trade(**patched_data)
        return [patched if t.id == trade.id else t for t in trades]
    return [t for t in trades if t.id != trade.id]


def validate_mutation(
    trades: list[Trade],
    mutation_type: MutationType,
    trade: Trade,
    patch: dict | None = None,
) -> tuple[bool, str, list[str]]:
    """
    가상 적용 후 oversell 여부 검증.

    Returns:
        (ok, message, affected_sell_ids)
    """
    virtual = _apply_virtual(trades, mutation_type, trade, patch)
    key = trade_to_group_key(trade)
    affected_sell_ids: list[str] = []

    for ev in walk_trades(
        virtual,
        group_filter=lambda t: is_same_group(t, key),
        sort_fn=sort_for_calc,
        track_fifo_lots=False,
    ):
        if ev.kind != "SELL":
            continue
        if ev.no_holding:
            return False, "보유 수량이 없어 매도할 수 없습니다.", []
        if ev.oversell:
            return False, "보유 수량이 부족한 매도 거래가 생깁니다.", []
        affected_sell_ids.append(ev.trade.id)

    return True, "", affected_sell_ids


def build_pnl_map(trades: list[Trade]) -> dict[str, float]:
    """저장된 profit_loss 값으로 SELL id → PnL 맵 구성."""
    return {t.id: float(t.profit_loss or 0) for t in trades if t.trade_type == TRADE_TYPE_SELL}
